test/replication-luatest/on_replication_split_brain_rollback_event_test.lua

```diff
@@ -0,0 +1,496 @@
+g.before_test('test_implicit_commit_failure', function(cg)
+    cg.master:exec(function()
+        local event = 'box.ctl.on_replication_split_brain_rollback'
+        _G.trigger.set(event, 't', function()
+            box.error.injection.set('ERRINJ_WAL_IO', true)
+            box.space.l:replace{0}
+            box.on_rollback(function()
+                box.error.injection.set('ERRINJ_WAL_IO', false)
+            end)
+        end)
+    end)
+end)
+
+-- Test that failure to implicitly commit a pending transaction after execution
+-- of the event triggers finishes caused split brain.
+g.test_implicit_commit_failure = function(cg)
+    t.tarantool.skip_if_not_debug()
+
+    cause_split_brain(cg.master, cg.replica)
+    cg.master:exec(function(replica_id)
+        t.helpers.retrying({timeout = 120}, function()
+            local upstream = box.info.replication[replica_id].upstream
+            t.assert_equals(upstream.status, 'stopped')
+            local msg = 'Failed to write to disk'
+            t.assert_equals(upstream.message, msg)
+        end)
+        t.assert_equals(box.info.synchro.queue.len, 1)
+    end, {cg.replica:get_instance_id()})
+end
+
+g.before_test('test_second_tx_failure', function(cg)
+    cg.master:exec(function()
+        local event = 'box.ctl.on_replication_split_brain_rollback'
+        _G.trigger.set(event, 't', function()
+            box.commit()
+            _G.trigger_ok = pcall(function() box.space.l:replace{0} end)
+        end)
+    end)
+end)
+
+-- Test that the second transaction started in the trigger will fail, since it
+-- won't have the `TXN_FORCE_ASYNC` flag set.
+g.test_second_tx_failure = function(cg)
+    cause_split_brain(cg.master, cg.replica)
+    cg.master:exec(function()
+        t.helpers.retrying({timeout = 120}, function()
+            t.assert_equals(box.info.synchro.queue.len, 0)
+        end)
+        t.assert_not(_G.trigger_ok)
+    end)
+end
+
+g.before_test('test_second_tx_implicit_commit_failure', function(cg)
+    cg.master:exec(function()
+        local event = 'box.ctl.on_replication_split_brain_rollback'
+        _G.trigger.set(event, 't', function()
+            box.commit()
+            box.begin()
+            box.space.l:replace{0}
+        end)
+    end)
+end)
+
+-- Test that the implicit commit of the second transaction started in the
+-- trigger will fail, since it won't have the `TXN_FORCE_ASYNC` flag set.
+g.test_second_tx_implicit_commit_failure = function(cg)
+    cause_split_brain(cg.master, cg.replica)
+    cg.master:exec(function(replica_id)
+        t.helpers.retrying({timeout = 120}, function()
+            local upstream = box.info.replication[replica_id].upstream
+            t.assert_equals(upstream.status, 'loading')
+            local msg = 'A rollback for a synchronous transaction is received'
+            t.assert_equals(upstream.message, msg)
+        end)
+        t.assert_equals(box.info.synchro.queue.len, 1)
+    end, {cg.replica:get_instance_id()})
+end
```

Comment:
I have to say - the test coverage is very impressive, I wouldn't even think about some of those cases myself.